modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts

```diff
@@ -1,22 +1,57 @@
+import { Menu, Toolbar } from '@ephox/bridge';
 import { Selections } from '@ephox/darwin';
 import { Arr, Cell, Optional, Thunk } from '@ephox/katamari';
-import { RunOperation, TableLookup } from '@ephox/snooker';
+import { RunOperation, Structs, TableLookup, Warehouse } from '@ephox/snooker';
 import { SugarElement, SugarNode } from '@ephox/sugar';
 import Editor from 'tinymce/core/api/Editor';
 import * as Util from '../core/Util';
 import * as TableTargets from '../queries/TableTargets';
 import * as TableSelection from './TableSelection';
 
-export type SelectionTargets = ReturnType<typeof getSelectionTargets>;
+export interface SelectionTargets {
+  readonly onSetupTable: (api: UiApi) => () => void;
+  readonly onSetupCellOrRow: (api: UiApi) => () => void;
+  readonly onSetupColumn: (lockedDisable: LockedDisable) => (api: UiApi) => () => void;
+  readonly onSetupPasteable: (getClipboardData: () => Optional<SugarElement[]>) => (api: UiApi) => () => void;
+  readonly onSetupPasteableColumn: (getClipboardData: () => Optional<SugarElement[]>, lockedDisable: LockedDisable) => (api: UiApi) => () => void;
+  readonly onSetupMergeable: (api: UiApi) => () => void;
+  readonly onSetupUnmergeable: (api: UiApi) => () => void;
+  readonly resetTargets: () => void;
+  readonly targets: () => Optional<RunOperation.CombinedTargets>;
+}
 
-export const getSelectionTargets = (editor: Editor, selections: Selections) => {
+interface ExtractedSelectionDetatils {
+  readonly mergeable: boolean;
+  readonly unmergeable: boolean;
+  readonly locked: Record<LockedDisableStrs, boolean>;
+}
+
+/*
+onAny - disable if any column in the selection is locked
+onFirst - disable if the first column in the table is selected and is locked
+onLast - disable if the the last column in the table is selected and is locked
+*/
+export const enum LockedDisable {
+  onAny = 'onAny',
+  onFirst = 'onFirst',
+  onLast = 'onLast'
+}
+
+type LockedDisableStrs = keyof typeof LockedDisable;
+
+type UiApi = Menu.MenuItemInstanceApi | Toolbar.ToolbarButtonInstanceApi;
```

Comment:
This should probably be above where you first use it, so it should be before the `SelectionTargets` interface.